packages/opentelemetry-resources/src/platform/node/detectors/AwsEcsDetector.ts

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Resource } from '../../../Resource';
+import { CONTAINER_RESOURCE } from '../../../constants';
+import { Detector } from '../../../types';
+import { ResourceDetectionConfigWithLogger } from '../../../config';
+import * as util from 'util';
+import * as fs from 'fs';
+import * as os from 'os';
+
+/**
+ * The AwsEcsDetector can be used to detect if a process is running in AWS
+ * ECS and return a {@link Resource} populated with data about the ECS
+ * plugins of AWS X-Ray. Returns an empty Resource if detection fails.
+ */
+export const cache = {
+  readFileAsync: util.promisify(fs.readFile),
+};
+
+class AwsEcsDetector implements Detector {
+  readonly CONTAINER_ID_LENGTH = 64;
+  readonly DEFAULT_CGROUP_PATH = '/proc/self/cgroup';
+  readonly ECS_METADATA_KEY_V4 = 'ECS_CONTAINER_METADATA_URI_V4';
+  readonly ECS_METADATA_KEY_V3 = 'ECS_CONTAINER_METADATA_URI';
+
+  async detect(config: ResourceDetectionConfigWithLogger): Promise<Resource> {
+    if (
+      !process.env.ECS_CONTAINER_METADATA_URI_V4 &&
+      !process.env.ECS_CONTAINER_METADATA_URI
+    ) {
+      config.logger.debug('AwsEcsDetector failed: Process is not on ECS');
+      return Resource.empty();
+    }
+
+    let containerId, hostName;
+    try {
+      hostName = os.hostname();
+    } catch (e) {
+      config.logger.warn(
+        `AwsEcsDetector failed to read host name: ${e.message}`
+      );
+    }
+
+    await this._getContainerId(config).then(res => {
+      containerId = res;
+    });
+
+    if (containerId && hostName) {
```

Comment:
Use a similar pattern to this with mutable `ResourceLabels` instead of branching every condition

https://github.com/open-telemetry/opentelemetry-js/blob/master/packages/opentelemetry-resources/src/platform/node/detectors/GcpDetector.ts#L56

Reply:
Thank you for pointing out! I also think this part is very dumb but not so sure how to write it in appropriate Typescript style.